### PR TITLE
[sui-core] Sign timestamp in headers

### PR DESCRIPTION
**Pull request by gdanezis**
We now include the header creation timestamp in the signed header data, so that we may extract a timestamp upon a commit. Signing a header is conditional on its creation time being in the past, ensuring that the created_at in all headers is in the past. in case the header is not in the past we wait for a short time, or return an error.

### CODE REVIEW

**Review by kchalkias — COMMENTED** [review]
While checking this PR I noticed that in messages.rs we use `to_be_bytes` see for `tracking_id` https://github.com/MystenLabs/sui/blob/main/crates/sui-types/src/messages.rs#L2024

What's our policy re `le` vs `be`? It seems that sticking to `le` is preferable (as I couldn't find any other uses of `be` at least in our Sui repo)

File: narwhal/types/src/primary.rs [hunk 1]
```diff
@@ -61,24 +61,24 @@ fn now() -> TimestampMs {
...
 // should not be treated as trustworthy data and should be used
 // for NON CRITICAL purposes only. For example should not be used
 // for any processes that are part of our protocol that can affect
 // safety or liveness.
-#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq, Arbitrary, MallocSizeOf)]
+//
+// However selected fields, such as
+// - Header::metadata::created_at
+// are signed to support commit timestamps.
+#[derive(Clone, Default, Serialize, Deserialize, Debug, PartialEq, Eq, Arbitrary, MallocSizeOf)]
```
Comment (andll) [inline narwhal/types/src/primary.rs:75]: Initially intention for this Metadata struct was to put fields that are not covered by signatures.

I feel like right now this Metadata struct might be too confusing in terms of what is signed and what is not.

Should we just take following approach:
- Remove the struct
- Add field `created_at` to Header
- Add field `created_at_untrusted` to Batch? (or, alternatively just include created_at into Batch digest as well even though it is not super needed)

**Review by andll — APPROVED** [review]
Overall this looks good, I added some comments about whether we should keep Metadata structure, but up to you

**Comment by andll** [comment]
For the `le` vs `be` policy - I think it is natural to use `le` everywhere, although it is not super critical for tracking_id, since it is not part of the core protocol

File: narwhal/types/src/primary.rs (hunk 1 above)
Comment (LefKok) [reply narwhal/types/src/primary.rs:75]: Even the Header timestamp is untrusted in the current design, the only think we can trust is the median of parent timestamps right?

Comment (andll) [reply narwhal/types/src/primary.rs:75]: Yeah, but original intention for Metadata struct was not whether it is trusted or not, but the fact whether it is covered by hash/signature or not.

**Pull request merged**